File: src/main/java/mc/rpgstats/component/StatsComponent.java
package mc.rpgstats.component;

import dev.onyxstudios.cca.api.v3.component.Component;
import dev.onyxstudios.cca.api.v3.component.sync.AutoSyncedComponent;
import mc.rpgstats.main.RPGStats;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.util.Identifier;

import java.util.HashMap;
import java.util.Objects;

public class StatsComponent implements Component, AutoSyncedComponent {
    private final PlayerEntity playerEntity;
    
    public HashMap<Identifier, StatsEntry> entries = new HashMap<>();
    
    public StatsComponent(PlayerEntity playerEntity) {
        this.playerEntity = playerEntity;
    }
    
    @Override
    public boolean shouldSyncWith(ServerPlayerEntity player) {
        return ServerPlayNetworking.canSend(player, RPGStats.SYNC_STATS_PACKET_ID);
    }
    
    @Override
    public void readFromNbt(NbtCompound compoundTag) {
        entries.clear();
        compoundTag.getKeys().forEach(s -> {
            Identifier identifier = Identifier.tryParse(s);
            if (identifier != null) {
                NbtCompound data = compoundTag.getCompound(identifier.toString());
                assert data != null;
                entries.put(identifier, new StatsEntry(identifier, data.getInt("level"), data.getInt("xp")));
            } else {
                System.err.println("Failed to parse stat identifier: " + s);
            }
        });
    }
    
    @Override
    public void writeToNbt(NbtCompound compoundTag) {
        for (StatsEntry entry : entries.values()) {
            entry.toCompound(compoundTag);
        }
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        StatsComponent that = (StatsComponent)o;
        return playerEntity.equals(that.playerEntity) && entries.equals(that.entries);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(playerEntity, entries);
    }
    
    public StatsEntry getOrCreateID(Identifier id) {
        return entries.computeIfAbsent(id, identifier -> new StatsEntry(id, 0, 0));
    }
}

File: src/main/kotlin/io/github/silverandro/rpgstats/stats/StatsComponent.kt
/*
 *   This Source Code Form is subject to the terms of the Mozilla Public
 *   License, v. 2.0. If a copy of the MPL was not distributed with this
 *   file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package io.github.silverandro.rpgstats.stats

import dev.onyxstudios.cca.api.v3.component.Component
import dev.onyxstudios.cca.api.v3.component.sync.AutoSyncedComponent
import io.github.silverandro.rpgstats.Constants
import io.github.silverandro.rpgstats.Constants.SYNC_STATS_PACKET_ID
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.nbt.NbtCompound
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.Identifier
import org.quiltmc.qsl.networking.api.ServerPlayNetworking
import java.util.*
import java.util.function.Consumer

class StatsComponent(private val playerEntity: PlayerEntity) : Component, AutoSyncedComponent {
    var entries = HashMap<Identifier, StatComponentEntry>()
    override fun shouldSyncWith(player: ServerPlayerEntity): Boolean {
        return ServerPlayNetworking.canSend(player, SYNC_STATS_PACKET_ID)
    }

    override fun readFromNbt(compoundTag: NbtCompound) {
        entries.clear()
        compoundTag.keys.forEach(Consumer { s: String ->
            val identifier = Identifier.tryParse(s)
            if (identifier != null) {
                val data = compoundTag.getCompound(identifier.toString())!!
                entries[identifier] = StatComponentEntry(identifier, data.getInt("level"), data.getInt("xp"))
            } else {
                Constants.LOG.error("Failed to parse stat identifier: $s")
            }
        })
    }

    override fun writeToNbt(compoundTag: NbtCompound) {
        for (entry in entries.values) {
            entry.addToCompound(compoundTag)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other!!.javaClass) return false
        val that = other as StatsComponent?
        return playerEntity == that!!.playerEntity && entries == that.entries
    }

    override fun hashCode(): Int {
        return Objects.hash(playerEntity, entries)
    }

    fun getOrCreateID(id: Identifier): StatComponentEntry {
        return entries.computeIfAbsent(id) { StatComponentEntry(id, 0, 0) }
    }
}